text/0693-property-injection.md

```diff
@@ -0,0 +1,526 @@
+We also recognize that there are situations where the development team might need to override the standard.
+In these situations, we value the autonomy of the development team by giving them the flexibility to override.
+
+In the Working Backwards example below, the Org defines standard defaults for development teams to use,
+but development teams can override the defaults if they need to.
+Development teams can also write their own Injectors and share them across teams.  
+
+## Working Backwards
+
+### Development Team Experience
+
+Development teams can start using Blueprints by attaching propertyInjectors to App.
+In the example below, this dev team wants use the Property Injectors provided by its org for S3 Buckets and Lambda Functions.
+
+```ts
+import { BucketPropsInjector, FunctionPropsInjector } from '@my-org/standard-props-injectors';
+...
+const app = new App({
+  ...
+  propertyInjectors: [
+    new BucketPropsInjector(),
+    new FunctionPropsInjector(),
+  ],
+});
+```
+
+Alternatively, you can use this syntax:
+
+```ts
+const app = new App({});
+
+PropertyInjectors.of(app).add(
+  new BucketPropsInjector(),
+  new FunctionPropsInjector(),
+);
+```
+
+Dev teams can also attach propertyInjectors to Stacks as well.  This will override the Bucket Injector provided in App.
+
+```ts
+import { BetaBucketPropsInjector } from '@my-team/custom-props-injectors';
+...
+const betaStack = new Stack(stage, 'MyApp-beta-stack', {
+  env: { account: '123456789012', region: 'us-east-1' },
+  propertyInjectors: [
+    new BetaBucketPropsInjector(),
+  ],
+});
+```
+
+In this case, Constructs created in betaStack will use `BetaBucketPropsInjector` instead of `BucketPropsInjector`.
+This is useful when Beta resources have different requirements.
+
+See [Scope Tree Traversal](#scope-tree-traversal) for more information.
+
+### Organization Level Standardization
+
+As an org with many dev teams, we can standardize how AWS Resources are created by providing dev teams with Injectors.
+
+Here is a simple PropertyInjector to make sure the S3 Bucket is not publically accessible and SSL is used.
+
+```ts
+class MyBucketPropsInjector implements IPropertyInjector {
+  public readonly constructUniqueId: string;
+
+  constructor() {
+    this.constructUniqueId = Bucket.PROPERTY_INJECTION_ID;
+  }
+
+  public inject(originalProps: BucketProps, _context: InjectionContext): BucketProps {
+    return {
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+      enforceSSL: true,
+      ...originalProps,
+    };
+  }
+}
+```
+
+Notes:
+
+* We will add Bucket.PROPERTY_INJECTION_ID to unique identify this Construct.
+This is a new Property that we will add to Bucket and other supported Constructs.
+* This TypeScript code will set blockPublicAccess to BlockPublicAccess.BLOCK_ALL and enforceSSL to true
+if these properties are not specified in originalProps.
```

Comment:
What happens if the properties are specified? It would be great to have a warning in case both the code and injector define a value for the property. Perhaps also the ability to explicitly suppress the warning.

Reply:
Do keep in mind that the current design will not make it possible for the property injector to add warnings *on* the construct (because when the injector executes, the construct won't exist yet!), so the best it can do is add warnings on its parent scope.